lib/types.js

```diff
@@ -1,35 +1,173 @@
 // @flow
 
-//
-// writing classic async JS with flow typing looks pretty ugly and is hard to read when defining the
-// types inline. So I figured I'd consolidate some common ones here that follow the cb(err,res) model.
-//
-// For example, instead of these hard-to-read beasts:
-//
-//     function a (cb : (err: ?Error, res: ?string) => void) : void {
-//     function b (cb : (err: ?Error) => void) : void {
-//     function c (arg: string, err : ?Error, res : ?{username: string, devicename: string}) => void) : void {
-//
-// we can write these:
-//
-//     function a (cb: CbString) : void {
-//     function b (cb: CbError)  : void {
-//     function c (cb: CbDeviceUsernamePair) : void {
-
-type DeviceUsernamePair = {username: string, devicename: string}
-type ChatChannel = {name: string, public: boolean, topic_type: string}
-type ChatSendMessage = {body: string}
-type ChatReadMessage = any
-
-type CbError = (err: ?Error) => void
-type CbAny = (err: ?Error, res: any) => void
-type CbDeviceUsernamePair = (err: ?Error, res: ?DeviceUsernamePair) => void
-
-type PromiseExec = Promise<any>
-
-type MessagesHandler = (arg: {messages: Array<ChatReadMessage>, channel: ChatChannel}) => void
-
-export type {DeviceUsernamePair, ChatChannel, ChatSendMessage, ChatReadMessage}
-export type {MessagesHandler}
-export type {CbAny, CbError, CbDeviceUsernamePair}
-export type {PromiseExec}
+/*
+ * Keybase Chat JSON API V1 options
+ * Source: https://github.com/keybase/client/blob/master/go/client/chat_api_handler.go
+ */
+
+export type ChatChannel = {
+  name: string,
+  public: boolean,
+  members_type: string,
+  topic_type?: string,
+  topic_name?: string,
+}
+
+export type ChatMessage = {
+  body: string,
+}
+
+export type Pagination = {
+  num: number,
+  next?: string,
+  previous?: string,
+  last?: string,
+}
+
+type ChatOptionsSearch = {
+  max_hits: number,
+  sent_by: string,
+  sent_after?: string,
+  sent_before?: string,
+  before_contex?: number,
+  after_context?: number,
+}
+
+/*
+ * List
+ * {"method": "list", "params": {"options": { ...ChatOptionsList }}}
+ */
+export type ChatOptionsList = {
```

Comment:
It might make more sense to make most of these types exact object types

Reply:
Agreed!